content/articles/understanding-xml-external-entity-attacks/index.md

```diff
@@ -0,0 +1,153 @@
+
+#### Document Type Definition (DTD)
+The DTD contains a definition of the structure and legal attributes and elements of a XMl document.DTD can be both within the document or loaded from an external source. When loading from an external source, we use the keyword `SYSTEM`. The external source can either be a URL or a file.
+
+A sample XML document with both DTD and entities will look like this.
+
+```xml
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE mail [ <!ENTITY sender “felix”> ]>
+<Greeting>
+      <From>&sender;</From>
+      <To>Section reader</To>
+      <Message>Hello, How are you today?</Message>
+</Greeting>
+```
+
+When the above code is parsed, we get this:
+
+```
+From: Felix
+
+To: Section reader
+
+Message: Hello, How are you today?
+```
+
+Now that we have learned how a basic XML document looks like, let’s learn how the attack works.
+
+### How XXE attacks work
+
+XXE attacks arise when external entities are supported and parsed by a weak XML parser. An attacker intercepts the XML data when in transit and adds malicious code. When processed, the application may disclose private information. Mostly these attacks enable the attackers to view the filesystem and, sometimes, they can interact with any back-end services that the application can access.
+
+There are several types of XXE attacks, such as:
+
+- **In-band XXE**: In this type, the attacker can receive an immediate response for the XXE payload they have uploaded.
+- **Out-of-band XXE (blind XXE)**: The attacker does not get an immediate response to their payload. To get a response, they reflect the output to another file or their server.
+
+[Portswigger](www.portswigger.net/web-security/xxe) has several labs that can help you practice. For this article, I will use one lab to show how XXE attacks can retrieve files.
+
+Using Burpsuite, we intercept the data and send it to the repeater.
+
+![intercept](/engineering-education/understanding-xml-external-entity-attacks/intercept1.png)
+
+Since the current data does not have any entities, we will introduce our own and add malicious code. Using the entities we introduce, we will try to check whether we can retrieve the password file. 
+
+![payload](/engineering-education/understanding-xml-external-entity-attacks/payload.png)
+
+If we are successful, we should be able to see the file as shown below:
+![passwords](/engineering-education/understanding-xml-external-entity-attacks/intercept2.png)
+
+Apart from retrieving files, we can use XXE attacks for :
+#### 1. Denial of Service (DoS)
+Most common being the billion laugh attack.
+
+#### 2. Server-side request forgery attacks
+Using the external entities, an attacker can make HTTP requests to URLs, then the server can access them, including those meant to be for internal use only in an organization.
+
+The attacker just places the URL that they are targeting. When the target is parsed, the contents of the said domain are exposed.
+
```

Comment:
The statement `The attacker just places the URL that they are targeting` is kind of hanging. Add more detail to this statement. eg. Where. For example, `The attacker just places the URL between ...`